### PR TITLE
Fix exponential execution time by memoizing format_expr

### LINKED ISSUE #4476: rustfmt is slow on pest generated grammars
Rustfmt takes about a minute to reformat 26k file (generated pest grammar).

File in question: https://github.com/steveklabnik/semver-parser/blob/0c43dbb12d35f9b58142b3976391eb52ea71725f/src/generated.rs

**To Reproduce**

Run this test in the semver-parser repo:

https://github.com/steveklabnik/semver-parser/blob/0c43dbb12d35f9b58142b3976391eb52ea71725f/tests/genpest.rs#L14

**Expected behavior**

Formatting takes low-digits seconds.

**Meta**

- rustfmt version: rustfmt 1.4.20-stable (48f6c32 2020-08-09)
- From where did you install rustfmt?: rustup
- How do you run rustfmt: https://github.com/steveklabnik/semver-parser/blob/0c43dbb12d35f9b58142b3976391eb52ea71725f/tests/genpest.rs#L68-L84

### LINKED ISSUE #4867: rustfmt takes 32 seconds for this 1140 character file
Formatting the code below takes 32 seconds on my machine.
This is not bad performance, this is a bug (If I remove one mod layer it takes 0.2 seconds).

In my real world, program (that uses pest), running `cargo expand` is therefore nearly impossible (I stopped it after an hour)

```
mod modA {
    mod modB {
        mod modC {
            mod modD {
                mod modE {
                    fn func() {
                        state . rule (Rule :: myrule , | state | { state . sequence (| state | { state . sequence (| state | { state . match_string ("abc") . and_then (| state | { super :: hidden :: skip (state) }) . and_then (| state | { state . match_string ("def") }) }) . and_then (| state | { super :: hidden :: skip (state) }) . and_then (| state | { state . sequence (| state | { state . optional (| state | { state . sequence (| state | { state . match_string ("abc") . and_then (| state | { super :: hidden :: skip (state) }) . and_then (| state | { state . match_string ("def") }) }) . and_then (| state | { state . repeat (| state | { state . sequence (| state | { super :: hidden :: skip (state) . and_then (| state | { state . sequence (| state | { state . match_string ("abc") . and_then (| state | { super :: hidden :: skip (state) }) . and_then (| state | { state . match_string ("def") }) }) }) }) }) }) }) }) }) }) });
                    }
                }
            }
        }
    }
}
```

### LINKED ISSUE #5128: rustfmt is prohibitively slow for sufficiently nested structs
The following snippet is about 220 lines containing a total of 13570 characters.

Running rustfmt against this file takes an inordinately long amount of time, with the following output from `--verbose` on
rustfmt 1.4.38-nightly (0fb1c371 2021-12-06)
`Spent 0.020 secs in the parsing phase, and 444.417 secs in the formatting phase`

This is actually only a substruct of the greater struct which I'm trying to validate. For the complete struct that I'm trying to validate, rustfmt exceeds 10 minutes of runtime, which makes it fatally time-out in my CI.

For context, these symbols come from Prost generated bindings against the `https://github.com/pganalyze/libpg_query` protobufs.

Is there known to be super-linear runtime scaling w.r.t. the amount of nesting of a struct or code block? This seems to be so slow that it ought to be considered a bug.

See also: https://github.com/rust-lang/rustfmt/issues/4867 

```
    use libpgquery_sys::proto::{
        self, node, AConst, AExpr, AStar, BoolExpr, ColumnRef, CommonTableExpr, FuncCall, Node,
        RangeVar, RawStmt, ResTarget, SelectStmt, SortBy, TypeCast, TypeName, WindowDef,
        WithClause,
    };

    fn takes_a_long_time_to_rustfmt() {
        let inner_cte = vec![Node {
            node: Some(node::Node::CommonTableExpr(Box::new(CommonTableExpr {
                ctename: String::from("ranked_by_age_within_key"),
                aliascolnames: vec![],
                ctematerialized: libpgquery_sys::proto::CteMaterialize::Default as i32,
                ctequery: Some(Box::new(Node { node: None })),
                location: 29,
                cterecursive: false,
                cterefcount: 0,
                ctecolnames: vec![],
                ctecoltypes: vec![],
                ctecoltypmods: vec![],
                ctecolcollations: vec![],
            }))),
        }];
        let outer_cte = vec![Node {
            node: Some(node::Node::CommonTableExpr(Box::new(CommonTableExpr {
                ctename: String::from("table_name"),
                aliascolnames: vec![],
                ctematerialized: libpgquery_sys::proto::CteMaterialize::Default as i32,
                ctequery: Some(Box::new(Node {
                    node: Some(node::Node::SelectStmt(Box::new(SelectStmt {
                        distinct_clause: vec![],
                        into_clause: None,
                        target_list: vec![
                            Node {
                                node: Some(node::Node::ResTarget(Box::new(ResTarget {
                                    name: String::from("column1"),
                                    indirection: vec![],
                                    val: Some(Box::new(Node {
                                        node: Some(node::Node::ColumnRef(ColumnRef {
                                            fields: vec![Node {
                                                node: Some(node::Node::String(proto::String {
                                                    str: String::from("c1"),
                                                })),
                                            }],
                                            location: 301,
                                        })),
                                    })),
                                    location: 301,
                                }))),
                            },
                            Node {
                                node: Some(node::Node::ResTarget(Box::new(ResTarget {
                                    name: String::from("column2"),
                                    indirection: vec![],
                                    val: Some(Box::new(Node {
                                        node: Some(node::Node::ColumnRef(ColumnRef {
                                            fields: vec![Node {
                                                node: Some(node::Node::String(proto::String {
                                                    str: String::from("c2"),
                                                })),
                                            }],
                                            location: 324,
                                        })),
                                    })),
                                    location: 324,
                                }))),
                            },
                        ],
                        from_clause: vec![Node {
                            node: Some(node::Node::RangeVar(RangeVar {
                                catalogname: String::from(""),
                                schemaname: String::from(""),
                                relname: String::from("ranked_by_age_within_key"),
                                inh: true,
                                relpersistence: String::from("p"),
                                alias: None,
                                location: 347,
                            })),
                        }],
                        where_clause: Some(Box::new(Node {
                            node: Some(node::Node::BoolExpr(Box::new(BoolExpr {
                                xpr: None,
                                boolop: libpgquery_sys::proto::BoolExprType::AndExpr as i32,
                                args: vec![
                                    Node {
                                        node: Some(node::Node::AExpr(Box::new(AExpr {
                                            kind: proto::AExprKind::AexprOp as i32,
                                            name: vec![Node {
                                                node: Some(node::Node::String(proto::String {
                                                    str: String::from("="),
                                                })),
                                            }],
                                            lexpr: Some(Box::new(Node {
                                                node: Some(node::Node::ColumnRef(ColumnRef {
                                                    fields: vec![Node {
                                                        node: Some(node::Node::String(
                                                            proto::String {
                                                                str: String::from("rank_in_key"),
                                                            },
                                                        )),
                                                    }],
                                                    location: 382,
                                                })),
                                            })),
                                            rexpr: Some(Box::new(Node {
                                                node: Some(node::Node::AConst(Box::new(AConst {
                                                    val: Some(Box::new(Node {
                                                        node: Some(node::Node::Integer(
                                                            proto::Integer { ival: 1 },
                                                        )),
                                                    })),
                                                    location: 396,
                                                }))),
                                            })),
                                            location: 394,
                                        }))),
                                    },
                                    Node {
                                        node: Some(node::Node::AExpr(Box::new(AExpr {
                                            kind: proto::AExprKind::AexprOp as i32,
                                            name: vec![Node {
                                                node: Some(node::Node::String(proto::String {
                                                    str: String::from("="),
                                                })),
                                            }],
                                            lexpr: Some(Box::new(Node {
                                                node: Some(node::Node::ColumnRef(ColumnRef {
                                                    fields: vec![Node {
                                                        node: Some(node::Node::String(
                                                            proto::String {
                                                                str: String::from("is_deleted"),
                                                            },
                                                        )),
                                                    }],
                                                    location: 402,
                                                })),
                                            })),
                                            rexpr: Some(Box::new(Node {
                                                node: Some(node::Node::TypeCast(Box::new(
                                                    TypeCast {
                                                        arg: Some(Box::new(Node {
                                                            node: Some(node::Node::AConst(
                                                                Box::new(AConst {
                                                                    val: Some(Box::new(Node {
                                                                        node: Some(
                                                                            node::Node::String(
                                                                                proto::String {
                                                                                    str:
                                                                                        String::from(
                                                                                            "f",
                                                                                        ),
                                                                                },
                                                                            ),
                                                                        ),
                                                                    })),
                                                                    location: 415,
                                                                }),
                                                            )),
                                                        })),
                                                        type_name: Some(TypeName {
                                                            names: vec![
                                                                Node {
                                                                    node: Some(node::Node::String(
                                                                        proto::String {
                                                                            str: String::from(
                                                                                "pg_catalog",
                                                                            ),
                                                                        },
                                                                    )),
                                                                },
                                                                Node {
                                                                    node: Some(node::Node::String(
                                                                        proto::String {
                                                                            str: String::from(
                                                                                "bool",
                                                                            ),
                                                                        },
                                                                    )),
                                                                },
                                                            ],
                                                            type_oid: 0,
                                                            setof: false,
                                                            pct_type: false,
                                                            typmods: vec![],
                                                            typemod: -1,
                                                            array_bounds: vec![],
                                                            location: -1,
                                                        }),
                                                        location: -1,
                                                    },
                                                ))),
                                            })),
                                            location: 413,
                                        }))),
                                    },
                                ],
                                location: 398,
                            }))),
                        })),
                        group_clause: vec![],
                        having_clause: None,
                        window_clause: vec![],
                        values_lists: vec![],
                        sort_clause: vec![],
                        limit_offset: None,
                        limit_count: None,
                        limit_option: libpgquery_sys::proto::LimitOption::Default as i32,
                        locking_clause: vec![],
                        with_clause: Some(WithClause {
                            ctes: inner_cte,
                            recursive: false,
                            location: 24,
                        }),
                        op: libpgquery_sys::proto::SetOperation::SetopNone as i32,
                        all: false,
                        larg: None,
                        rarg: None,
                    }))),
                })),
                location: 5,
                cterecursive: false,
                cterefcount: 0,
                ctecolnames: vec![],
                ctecoltypes: vec![],
                ctecoltypmods: vec![],
                ctecolcollations: vec![],
            }))),
        }];
    }
```

### PR DESCRIPTION
**Pull request by HKalbasi**
fix #4476 
fix #4867
fix #5128 

The problem in those issues is, when `max_width` is tight, rustfmt should check all possible formattings, which is exponential relative to input size. By [memoization](https://en.wikipedia.org/wiki/Memoization) of `format_expr` function, worst case execution time becomes `O(n*max_width^3)`. I tried codes in those issues and it completed in a fraction of second.

### CODE REVIEW

**Comment by HKalbasi** [comment]
Test are failing strangely. I think my `QueryId` is not unique enough. What is needed to be added?

**Comment by calebcartwright** [comment]
> Test are failing strangely. I think my `QueryId` is not unique enough. What is needed to be added?

I can't say definitively whether there actually is a combination that will prove sufficiently unique, so you'll probably have to continue experimenting to gauge feasibility. Perhaps the expression's nodeid?

I also wouldn't be terribly keen on introducing a mutexted global, would it be feasible to attach the map to the rewrite context somehow?

**Comment by HKalbasi** [comment]
> I can't say definitively whether there actually is a combination that will prove sufficiently unique, so you'll probably have to continue experimenting to gauge feasibility. Perhaps the expression's nodeid?

Expression's node id is shared between all expressions of a def id. But I found that if I exclude macros (memoization doesn't help there anyway) it will pass tests. Probably span value is unique outside of macros.

> I also wouldn't be terribly keen on introducing a mutexted global, would it be feasible to attach the map to the rewrite context somehow?

I tried adding a `RefCell` to rewrite context, (last commit) but it doesn't improve performance and those issues still takes minutes. But the very same code with global mutex works. I don't know what is happening.

**Comment by calebcartwright** [comment]
> I tried adding a RefCell to rewrite context, (last commit) but it doesn't improve performance and those issues still takes minutes. But the very same code with global mutex works. I don't know what is happening.

Sorry haven't had a chance to dig into this in detail and probably won't for the next couple weeks due to the holidays. I'd suspect that's a case of a cache miss with the former, perhaps due to some slightly different logic needed vs. the global.

Overall I think you're on to something here so really appreciate you digging in! Would like to continue iterating on the implementation specifics though, as well as incorporating some tests

**Comment by HKalbasi** [comment]
The problem was that rewrite context doesn't live long enough, so I moved it one layer up and it is now fixed.

> as well as incorporating some tests

I added examples in those issues as regular tests, so a regression would cause `cargo test` to not terminate in reasonable time (but it won't fail). A global timeout for all tests would make it to fail.

**Comment by calebcartwright** [comment]
Fantastic thank you!

File: src/rewrite.rs [hunk 1]
```diff
@@ -24,10 +26,19 @@ impl<T: Rewrite> Rewrite for ptr::P<T> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub(crate) struct QueryId {
+    pub(crate) shape: Shape,
+    pub(crate) span: SpanData,
+}
+
+pub(crate) type Memoize = Option<HashMap<QueryId, Option<String>>>;
```
Comment (camsteffen) [inline src/rewrite.rs:35]: Is the outer `Option` necessary?

Comment (camsteffen) [inline src/rewrite.rs:35]: Maybe you can use `FxHashMap`?

File: src/rewrite.rs [hunk 2]
```diff
@@ -24,10 +26,19 @@ impl<T: Rewrite> Rewrite for ptr::P<T> {
...
+
 #[derive(Clone)]
 pub(crate) struct RewriteContext<'a> {
     pub(crate) parse_sess: &'a ParseSess,
     pub(crate) config: &'a Config,
+    pub(crate) memoize: Rc<RefCell<Memoize>>,
```
Comment (camsteffen) [inline src/rewrite.rs:41]: Can you use `Cell` instead of `RefCell`, similar to other fields here?

File: src/rewrite.rs [hunk 3]
```diff
@@ -24,10 +26,19 @@ impl<T: Rewrite> Rewrite for ptr::P<T> {
     }
 }
 
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub(crate) struct QueryId {
+    pub(crate) shape: Shape,
+    pub(crate) span: SpanData,
```
Comment (camsteffen) [inline src/rewrite.rs:32]: I think `Span` would be better here.

**Comment by camsteffen** [comment]
> Expression's node id is shared between all expressions of a def id.

Maybe this is because id's arent assigned until macro expansion ([code](https://github.com/rust-lang/rust/blob/2b67c30bfece00357d5fc09d99b49f21066f04ba/compiler/rustc_expand/src/expand.rs#L1626)), which I assume doesn't run for rustfmt? If that is the case, I wonder if it would be worthwhile to make a change upstream so that node ids may be assigned without macro expansion.

File: src/rewrite.rs (hunk 1 above)
Comment (HKalbasi) [reply src/rewrite.rs]: Thanks for reviewing! Yes, in case of `None` function should clear the store at the end, but in `Some(empty)` it should fill it and keep it.

File: src/rewrite.rs (hunk 2 above)
Comment (HKalbasi) [reply src/rewrite.rs:40]: It isn't possible to mutate things behind `Cell` in-place (insert to hashmap), I think.

File: src/rewrite.rs (hunk 1 above)
Comment (camsteffen) [reply src/rewrite.rs]: Hmm wouldn't it be the same to do `clean = map.is_empty()`?

File: src/rewrite.rs (hunk 2 above)
Comment (camsteffen) [reply src/rewrite.rs:40]: You can `take`, mutate and replace. More ceremony but I believe more efficient after optimizations.

Comment (HKalbasi) [reply src/rewrite.rs:40]: Hmm, so basically `RefCell<T>` is equivalent to `Cell<Option<T>>` with unwraping on `borrow_mut` and replacing on drop? Interesting. Can you explain why it would optimize better? About similarity to other fields, there is a `pub(crate) skipped_range: Rc<RefCell<Vec<(usize, usize)>>>` there as well.

Is it possible/better to remove all cells from rewrite context, pass it by `&mut` to most functions that currently get it by `&`, and wrap it in a `Rc<RefCell<` if/when we need to clone it? Certainly out of scope for this PR.

File: src/rewrite.rs (hunk 1 above)
Comment (HKalbasi) [reply src/rewrite.rs]:
```
*map = Some(HashMap::default());
clean = true;
```
If we set `clean = map.is_empty()`, then `*map = Some(HashMap::default());` should become inserting some dummy thing (which isn't good) to map, otherwise subsequent calls will also receive empty map and clean it, while they shouldn't.

File: src/rewrite.rs (hunk 2 above)
Comment (camsteffen) [reply src/rewrite.rs:40]: > Can you explain why it would optimize better?

RefCell enforces borrow checking rules at runtime which has a small cost.

File: src/rewrite.rs (hunk 1 above)
Comment (camsteffen) [reply src/rewrite.rs]: I see. I wonder if there is a cleaner solution here, such as clearing the map after each `Item` is formatted.

**Comment by calebcartwright** [comment]
Thanks again @HKalbasi! Few questions/asks:

- Could you please incorporate some inline comments explaining the why and thinking? Even after just a couple weeks I found I'd forgotten a lot of context, and I can only imagine someone trying to look at this for the first time months/years from now
- I feel like we should be able to avoid at least one of clones, most likely once we have a hit. Would using a mutable borrow on the map and `remove_entry` instead of `get` [here](https://github.com/rust-lang/rustfmt/pull/5139/files#diff-4858c8e24a51bcd35635e129cb086442007e34a4029f4333c6bacdfec94d155fR67-R70) do the trick? 
- Any thoughts on some of the additional optimizations @camsteffen suggested, e.g. https://github.com/rust-lang/rustfmt/pull/5139#discussion_r776736592 ?

File: src/rewrite.rs (hunk 2 above)
Comment (HKalbasi) [reply src/rewrite.rs]: Replaced with Cell

File: src/expr.rs [hunk 4]
```diff
@@ -53,6 +54,54 @@ pub(crate) fn format_expr(
...
+    }
+    let clean;
+    let query_id = QueryId {
+        shape,
+        span: expr.span,
+    };
+    if let Some(map) = context.memoize.take() {
+        if let Some(r) = map.get(&query_id) {
+            let r = r.clone();
+            context.memoize.set(Some(map)); // restore map in the memoize cell for other users
```
Comment (calebcartwright) [inline src/expr.rs:75]: This is the piece I'm struggling a bit to conjure up mentally. If we've got the final formatting for a specific expression, in what context would we need to get that particular entry again given the hierarchical structure/who would other users be?

**Comment by HKalbasi** [comment]
> I feel like we should be able to avoid at least one of clones, most likely once we have a hit. Would using a mutable borrow on the map and remove_entry instead of get here do the trick?

Nothing prevent it from hitting more than once. So if we remove it from map in the first hit, we need to calculate it again. If clones are too bad, I think we can use `Rc` since caller just read the result, but I don't feel it is necessary.

> Any thoughts on some of the additional optimizations @camsteffen suggested

I don't think that's an optimization, but it would make code cleaner. I'm not confident where to clean cache so it cleans enough and not more. And in current way all of codes related to memoizing are in the same function, so lets keep it as is. I tried to explain why there is `Option<HashMap>` in the comments.

The second one (RefCell vs Cell) was an optimization, and it's done.

File: src/expr.rs (hunk 4 above)
Comment (HKalbasi) [reply src/expr.rs:75]: In the recursive calls it will be called again if formating fails due exceeding max_width limit:
```
loooooooooooong(looooooooooong(this_one_formats(twice)))
```
calculate formatting of `this_one_formats(twice)` but it fails later, and we try this:
```
loooooooooooong(
    looooooooooong(
         this_one_formats(twice)
   )
)
```
this formats `this_one_formats(twice)` identical to above, but calculate it independently.

So other users here are other recursive calls to the `format_expr` functions.

**Comment by calebcartwright** [comment]
re: my question in https://github.com/rust-lang/rustfmt/pull/5139#discussion_r795093714 and :point_down: 
> Nothing prevent it from hitting more than once. So if we remove it from map in the first hit, we need to calculate it again.

I was trying to think of a scenario where we'd have to recompute a `Some(...)` with the exact same shape (and I still can't see that), but I think I could see a scenario where it's a `None` that we could potentially hit twice. Does that sound right or is there a scenario for the former you can think of?

In short, I'm trying to be mindful of time vs. space complexity, especially since the originating issues occur in only the most extreme of circumstances derived from generated code whereas this change is going to be impact quite literally everything. As such, I do want to be mindful of clones

File: src/expr.rs (hunk 4 above)
Comment (calebcartwright) [reply src/expr.rs:75]: discussion shifted outside this comment so resolving

**Comment by HKalbasi** [comment]
I'm not sure at this point if we hit `Some` in the cache. If there is no such case, we can use a `HashSet` instead of `HashMap` and return `None` if we calculated it before. But would it format correctly if this conjecture becomes wrong?

> In short, I'm trying to be mindful of time vs. space complexity, especially since the originating issues occur in only the most extreme of circumstances derived from generated code whereas this change is going to impact quite literally everything. As such, I do want to be mindful of clones. If there's no way around it then so be it, but it's certainly worth the thought exercise

Yes, now I see the second clone is pure overhead in normal cases. `Rc` is still an option.

File: tests/target/performance/issue-4476.rs [hunk 5]
```diff
@@ -0,0 +1,705 @@
...
+                                                                .optional(|state| {
+                                                                    state.match_string(",")
+                                                                })
+                                                                .and_then(|state| {
+                                                                    super::hidden::skip(state)
+                                                                })
+                                                                .and_then(|state| {
+                                                                    state.sequence(|state| {
+                                                                        self::space(state)
+                                      .and_then(|state| super::hidden::skip(state))
```
Comment (calebcartwright) [inline tests/target/performance/issue-4476.rs:192]: The indentation here is quite off. I can't say offhand whether that's directly related because we don't really have a good frame of reference (as IIRC rustfmt on master is crashing on this input), but it could be indicative of a shape issue. The human-level diff comparison is also tough to do by eye given the heavy ident reuse.

Would you be willing to try to boil this down to a smaller repo, starting with the containing function at the same indentation level to see if we can rule out there being any shape-related oddities in the map entries? If it turns out to just be a chain issue (which wouldn't surprise me in the least) that'll be good to know, but definitely doesn't need to be resolved here

**Comment by calebcartwright** [comment]
> I'm not sure at this point if we hit `Some` in the cache. ~If there is no such case, we can use a `HashSet` instead of `HashMap` and return `None` if we calculated it before. But would it format correctly if this conjecture becomes wrong?~ (Edit: I tried it and tests failed)

I won't speak to the set vs. map piece, though to the original question, the only way rustfmt would ever format one of these expressions _incorrectly_ would be if we'd done something egregious and were storing the wrong `Some(String...)` for the wrong query id. In practice the real worst case would be returning a `None` speciously, in which case rustfmt simply emits the original span contents of the expression (functionally skipping attempting to format that expression). Not ideal of course, but it's something done elsewhere on other code paths so certainly not disastrous either.

> Yes, now I see the second clone is pure overhead in normal cases. Rc is still an option.

Will differ to your preference, but I want to note the execution time numbers I'm seeing locally with your current version (clone) and with the adjustment I suggested to remove the entry from the map via a mutable ref so as to avoid the need for the clone in the return case:

(note I've elided the `--check` contents as it's irrelevant noise in this context):

Current with the `get` + clone:
```bash
$ time target/debug/rustfmt tests/target/performance/issue-5128.rs --check
.....

real	0m0.632s
user	0m0.619s
sys	0m0.012s
```

With `take_entry` skipping the clone:
```bash
$ time target/debug/rustfmt tests/target/performance/issue-5128.rs --check
.....

real	0m0.635s
user	0m0.622s
sys	0m0.012s
```

And using the original/source input, as-is with `get` + clone
```bash
$ time target/debug/rustfmt tests/source/performance/issue-5128.rs --check
.....

real	0m8.926s
user	0m8.822s
sys	0m0.069s
```

with `take_entry`:
```bash
$ time target/debug/rustfmt tests/source/performance/issue-5128.rs --check
.....

real	0m8.776s
user	0m8.734s
sys	0m0.041s
```

Realize the above doesn't exactly qualify as a statistically significant analysis but I'm seeing the same pattern consistently across multiple runs, and for the other input files as well. This gives me sufficient confidence the extra clone has no substantial execution time impact, at least not in any context known to us today, and we should go with one of the discussed approaches to avoid it.

**Comment by calebcartwright** [comment]
Also just want to say thanks again for all your efforts on this! I think it'll be a great step forward and address several known problems, but also want to be very careful and thoughtful given the code paths that are getting touched.

**Comment by HKalbasi** [comment]
Third hits are not impossible, we can see by testing this (it fails):
```rust
if let Some(r) = map.remove(&query_id) {
    map.insert(query_id, Some("we don't hit again".to_string()));
    context.memoize.set(Some(map)); // restore map in the memoize cell for other users
    return r;
}
```
Your numbers shows that things which we calculate exactly two times are more than things which we calculate more times. Existence of third hits can lead to exponential time theoretically, thought it isn't probably important in practical cases. Anyway, that clone is in cold path in normal codes (`issue-5128.rs` isn't normal code), and what we did before memoizing was to completely recalculate the string from the ground up, which is definitely more costly than cloning, and it was fine.

What clone that is IMO more important, and has overhead relative to previous state for normal codes, is the second clone:
```rust
if let Some(mut map) = context.memoize.take() {
    map.insert(query_id, r.clone()); // insert the result in the memoize map
    context.memoize.set(Some(map)); // so it won't be computed again
}
```
In normal cases, most of expressions won't even hit once and they will waste. I don't have solution for it. I tried `Rc` and it doesn't work either because some (most?) of the callers want to mutate the result (for example append something) so they need to clone the actual data anyway.

> Also just want to say thanks again for all your efforts on this! I think it'll be a great step forward and address several known problems, but also want to be very careful and thoughtful given the code paths that are getting touched.

I completely understand your concerns, and thanks for maintaining rustfmt!

**Comment by calebcartwright** [comment]
> What clone that is IMO more important, and has overhead relative to previous state for normal codes, is the second clone

Yes, though I'd basically accepted that we're going to have to introduce at least one, just as one of the inherent drawbacks with memoization. While I agree that the second clone won't always be at play in normal (non-generated, non-borderline-pathlogical :laughing:) code, I feel like it may with certain expressions in scenarios such as RHS comparisons.

> Third hits are not impossible, we can see by testing this (it fails):

That's a much more practical way of testing it than my trying to mentally execute code flows!

Based on what we're seeing here I'm starting to suspect there's probably a middle ground somewhere that would both be a marked improvement in the generated code cases while also still minimizing unnecessary clones in more standard cases, even if the result is still less than optimal from an asymptotic complexity pov.

I wonder if we could use a more advanced type for the map values which included some kind of a hit counter that would be used to conditionally determine whether to clone + attach the result to the map entry, and potentially conditionally remove the entry as well to split the difference on some of the cloning I was previously referring to

**Comment by HKalbasi** [comment]
I measured execution time of system and idempotence tests (I know they are not necessarily good data, just it was easy to add a timer in `run_test_with`) in some scenarios, excluding tests in this PR. Results have a ±0.03 second error, I tried to pick the mean in multiple runs.

1. Without caching:
```
Time elapsed in system: 0.579 seconds
Time elapsed in idempotence: 0.704 seconds
```
2. This PR at current state: Seems it is helping in normal code as well!
```
Time elapsed in system: 0.531 seconds
Time elapsed in idempotence: 0.640 seconds
```
3. Removing from map at first hit:
```
Time elapsed in system: 0.544 seconds
Time elapsed in idempotence: 0.632 seconds
```
results are mixed, but I think 3 is slightly faster. The difference is cost of the second clone, minus cost of recalculations in third times. 

I tried to make a version of 1 with an additional clone, but optimizer seems catch me and remove the clone.

> Based on what we're seeing here I'm starting to suspect there's probably a middle ground somewhere that would both be a marked improvement in the generated code cases while also still minimizing unnecessary clones in more standard cases, even if the result is still less than optimal from an asymptotic complexity pov.

I'm fine with removing from map on first hit, but data above doesn't support it makes a noticeable difference.

And I want to mention `Rc`/`Cow` again, using it needs some change in most of the codebase (so out of scope for this PR), but it probably saves the first clones in addition to the second clones.

**Comment by calebcartwright** [comment]
I'd put this on hold for a bit as the release we had pending was already large and I didn't want to try to pull this in on top, but am circling back to it now.

I think I've reached a point where I'm comfortable merging, though will probably wait a tad longer to do so as I'd like to do some additional experimentation shortly after and have the bandwidth to do that before we have too many other commits building on top of it.

I am, however, unsure as to what you were trying to test/assert with your last comment as it seems duplicative of what was established earlier relative to timing: [my prior point](https://github.com/rust-lang/rustfmt/pull/5139#issuecomment-1028577553) was that the run/execution time was largely the same regardless of whether there was an extra clone, so I'd rather avoid allocating increasingly long strings that weren't providing any benefit at runtime nor IMO pre-compile time in terms of the source code.

> And I want to mention Rc/Cow again, using it needs some change in most of the codebase (so out of scope for this PR), but it probably saves the first clones in addition to the second clones.

Apologies if I forgot to respond to this thread previously, but I saw it and agreed completely that I don't want to go down this route.

**Comment by calebcartwright** [comment]
Also, the review comment I left in https://github.com/rust-lang/rustfmt/pull/5139#discussion_r798192972 is still outstanding and is something we'd need clarity on before proceeding. I can't tell if the shape is just wrong or if we've perhaps got a mapping issue, but something is wrong there.

File: tests/target/performance/issue-4476.rs (hunk 5 above)
Comment (HKalbasi) [reply tests/target/performance/issue-4476.rs:192]: Any example should fill the `max_width` limit, either with indentation or with decreasing the limit. example in the #4867 is fairly minimal and I tried to minimize it more:
```rust
fn func() {
    state . rule (Rule :: myrule , | state | { state . sequence (| state | { state . sequence (| state | { state . match_string ("abc") . and_then (| state | { super :: hidden :: skip (state) }) . and_then (| state | { state . match_string ("def") }) }) . and_then (| state | { super :: hidden :: skip (state) }) . and_then (| state | { state . sequence (| state | { state . optional (| state | { state . sequence (| state | { state . match_string ("abc") . and_then (| state | { super :: hidden :: skip (state) }) . and_then (| state | { state . match_string ("def") }) }) . and_then (| state | { state . repeat (| state | { state . sequence (| state | { super :: hidden :: skip (state) . and_then (| state | { state . sequence (| state | { state . match_string ("abc") }) }) }) }) }) }) }) }) }) });
}
```
this will take 7s on my system with (`max_width=70`) and then return no solution. Examples that have solution in the `max_width` limit are faster since the solution is not the last thing that we try, so minimizing them is harder because each step halves the execution time. Does this example helps?

**Comment by HKalbasi** [comment]
> so I'd rather avoid allocating increasingly long strings that weren't providing any benefit at runtime nor IMO pre-compile time in terms of the source code.

I don't think its affect on memory usage would be more than ten megabytes even in machine generated codes (we can measure it) so is it that bad?

**Comment by calebcartwright** [comment]
> I don't think its affect on memory usage would be more than ten megabytes even in machine generated codes (we can measure it) so is it that bad?

This is an increasingly moot topic of debate because as noted previously the only outstanding issue needing resolution is relative to this change producing incorrect formatting.

I'm not requesting that you make any changes on this piece, however, I want to stress that we don't want to just incorporate unhelpful/unnecessary steps into the codebase unless they clear a threshold of being unnecessarily "bad enough"; we've very much a strategy of "only do that which is necessary".

There's plenty of cases, both those that could easily be contrived and practical ones within the rustfmt codebase, where some minimally (or even be negligibly) impactful step could be unnecessarily taken, e.g. a requiring clone instead of working with a slice. However, we'd still want to avoid such an unnecessary action regardless, i.e. go with the slice instead of the unnecessary clone.

I'm simply pointing out that the exact same principle is applicable here: we need to have a rationalization for the usage of a clone; we don't want to use clones freely/loosely and then have to rationalize the removal/avoidance of them

File: tests/target/performance/issue-4476.rs (hunk 5 above)
Comment (calebcartwright) [reply tests/target/performance/issue-4476.rs:192]: Sorry but I don't understand your response in relation to the question that was posed. The issue here is that the emitted formatting is wrong. The runtime performance of how long it takes to produce the incorrect formatting is orthogonal.

Do you have any insight into why this is producing an incorrect result?

Comment (HKalbasi) [reply tests/target/performance/issue-4476.rs:192]: Ah, I completely ignored the selected lines and misunderstood your message. This seems very bad, will look at it.

Comment (calebcartwright) [reply tests/target/performance/issue-4476.rs:192]: no worries, and thank you!

Comment (HKalbasi) [reply tests/target/performance/issue-4476.rs:192]: It seems in lines 192-207 tabs are 2 spaces but in rest tabs are 4 spaces, any ideas?

Comment (calebcartwright) [reply tests/target/performance/issue-4476.rs:192]: the input had two spaces iirc, so perhaps it's somehow "successfully" falling back to the span contents of the input? fwiw it might be more helpful on human eyes to butcher the input formatting a bit more and/or use some different idents given how much similarity there is (as that would make things stand out more between input/output)

Comment (HKalbasi) [reply tests/target/performance/issue-4476.rs:192]: You are right, if I change the spaces to 1, it will be 1 in the output. And I can reproduce it on rustfmt installed on my machine after ~50s, so it is not caused by this PR.

Comment (calebcartwright) [reply tests/target/performance/issue-4476.rs:192]: Okay thank you for confirming. That strikes me as very odd behavior considering that we're dealing with a chain (a massive one but a chain nevertheless) and how chain formatting behaves in less.. intense scenarios, but so long as it's not something being introduced/modified as part of this PR then we should be good.

Comment (HKalbasi) [reply tests/target/performance/issue-4476.rs:192]: Slightly minimized, takes 3s on rustfmt master: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d36a8d8e3ddc4b6b0223144a4b73ded9

Increasing `max_width` will fix the problem, so there is a possiblity that this problem have the same root with the long running time.

**Comment by calebcartwright** [comment]
Alrighty, let's give this a try!

**Comment by ytmimi** [comment]
@calebcartwright given that we just had to revert these changes in #5403 do we want to reopen all the linked issues?

**Comment by HKalbasi** [comment]
Ah, sorry for breaking rustfmt.

If I understand correctly, the problem is that the result is not a function of just `shape` and `expr`. Would including the fields of `RewriteContext` (except `config`, `parse_sess` and `snippet_provider` which it seems they are global) in `QueryId` be enough to solve this problem?

**Comment by calebcartwright** [comment]
No worries! This is a fairly tricky area and I didn't foresee this either.

I suspect there will be more to it and that we may need a more advanced/conditional type of memoization flow like I was alluding to earlier in these discussions (albeit with even more conditions)
